extensions/panache/hibernate-orm-panache-common/runtime/src/main/java/io/quarkus/hibernate/orm/panache/common/runtime/AbstractJpaOperations.java

```diff
@@ -421,7 +431,43 @@ public boolean exists(Class<?> entityClass, String query, Parameters params) {
     }
 
     public long deleteAll(Class<?> entityClass) {
-        return (long) getEntityManager(entityClass).createQuery("DELETE FROM " + getEntityName(entityClass)).executeUpdate();
+        //detecting the case where there are cascade-delete associations, and do the the bulk delete query otherwise.
+        if (deleteOnCascadeDetected(entityClass)) {
+            int count = 0;
+            List<?> objects = list(findAll(entityClass));
+            for (Object entity : objects) {
+                getEntityManager().remove(entity);
+                count++;
+            }
+            return count;
+        }
+        return (long) getEntityManager().createQuery("DELETE FROM " + getEntityName(entityClass)).executeUpdate();
+
+    }
+
+    /**
+     * Detects if cascading delete is needed. The delete-cascading is needed when associations with cascade delete enabled
+     * {@link javax.persistence.OneToMany#cascade()} and also on entities containing a collection of elements
+     * {@link javax.persistence.ElementCollection}
+     * 
+     * @param entityClass
+     * @return true if cascading delete is needed. False otherwise
+     */
+    private boolean deleteOnCascadeDetected(Class<?> entityClass) {
+        Metamodel metamodel = getEntityManager().getMetamodel();
+        EntityType<?> entity1 = metamodel.entity(entityClass);
+        Set<Attribute<?, ?>> declaredAttributes = ((EntityTypeImpl) entity1).getDeclaredAttributes();
+
+        CascadeStyle[] propertyCascadeStyles = getEntityManager().unwrap(SessionImplementor.class)
+                .getEntityPersister(entityClass.getName(), null)
+                .getPropertyCascadeStyles();
+        boolean doCascade = Arrays.stream(propertyCascadeStyles)
+                .filter(cascadeStyle -> cascadeStyle.doCascade(CascadingActions.DELETE))
+                .count() > 0;
+        boolean hasElementCollection = declaredAttributes.stream().filter(attribute -> attribute.getPersistentAttributeType()
+                .equals(Attribute.PersistentAttributeType.ELEMENT_COLLECTION)).count() > 0;
```

Comment:
You need to at least also detect `MANY_TO_MANY` here.

Reply:
Not sure to understand. I'm not detecting the kind of association (`MANY_TO_MANY` or `ONE_TO_MANY` and so on) but if the association has the `cascade` mode enabled. So the `MANY_TO_MANY` with cascade enabled will be detected here.
In the other hand, when an attribute has `@ElementCollection`, it is meaningless without the parent entity, so, whenever a parent entity is removed, the children will also be, no?

Reply:
The thing is that many-to-many associations have the link table that needs to be cleaned up when you delete the current entity, even if you're *not* cascading the delete to the associated entity (which you wouldn't usually do anyway for a many-to-many).

Reply:
(So you need to treat many-to-many the same as an element collection.)

Reply:
I have added a test to verify that many-to-many associations are correctly managed and it's ok, the link table is also cleaned up. I've just pushed the code.